### PR TITLE
de-null dartDefines in daemon mode

### LINKED ISSUE #45048: Flutter for Web is failing to run a release build
## Steps to Reproduce
Have tried on two different projects which are using the same Flutter SDK
1. Upgrade Flutter to the latest version
2. Try to run in release or profile mode

## Logs
```
Launching lib/main.dart on Web Server in release mode...
Building application for the web...
Compiling lib/main.dart for the Web...
#0      Object.noSuchMethod (dart:core-patch/object_patch.dart:53:5)
#1      parseDartDefines (package:flutter_tools/src/build_system/targets/dart.dart:388:26)
#2      Dart2JSTarget.build (package:flutter_tools/src/build_system/targets/web.dart:151:33)
#3      _BuildInstance._invokeInternal (package:flutter_tools/src/build_system/build_system.dart:526:25)
<asynchronous suspension>
#4      _BuildInstance.invokeTarget.<anonymous closure> (package:flutter_tools/src/build_system/build_system.dart:481:35)
#5      new Future.sync (dart:async/future.dart:224:31)
#6      AsyncMemoizer.runOnce (package:async/src/async_memoizer.dart:43:45)
#7      _BuildInstance.invokeTarget (package:flutter_tools/src/build_system/build_system.dart:481:21)
<asynchronous suspension>
<asynchronous suspension>
#8      BuildSystem.build (package:flutter_tools/src/build_system/build_system.dart:419:36)
#9      buildWeb (package:flutter_tools/src/web/compile.dart:41:50)
<asynchronous suspension>
#10     WebFs.start (package:flutter_tools/src/build_runner/web_fs.dart:309:13)
<asynchronous suspension>
#11     _DwdsResidentWebRunner.run.<anonymous closure> (package:flutter_tools/src/build_runner/resident_web_runner.dart:598:36)
#12     asyncGuard.<anonymous closure> (package:flutter_tools/src/base/async_guard.dart:111:32)
#13     _rootRun (dart:async/zone.dart:1126:13)
#14     _CustomZone.run (dart:async/zone.dart:1023:19)
#15     _runZoned (dart:async/zone.dart:1518:10)
#16     runZoned (dart:async/zone.dart:1502:12)
#17     asyncGuard (package:flutter_tools/src/base/async_guard.dart:109:3)
#18     _DwdsResidentWebRunner.run (package:flutter_tools/src/build_runner/resident_web_runner.dart:597:32)
<asynchronous suspension>
#19     AppDomain.startApp.<anonymous closure> (package:flutter_tools/src/commands/daemon.dart:470:23)
#20     AppDomain.launch.<anonymous closure> (package:flutter_tools/src/commands/daemon.dart:532:26)
#21     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:146:29)
#22     _rootRun (dart:async/zone.dart:1126:13)
#23     _CustomZone.run (dart:async/zone.dart:1023:19)
#24     _runZoned (dart:async/zone.dart:1518:10)
#25     runZoned (dart:async/zone.dart:1465:12)
#26     AppContext.run (package:flutter_tools/src/base/context.dart:145:18)
#27     AppInstance._runInZone (package:flutter_tools/src/commands/daemon.dart:930:20)
#28     AppDomain.launch (package:flutter_tools/src/commands/daemon.dart:530:15)
#29     AppDomain.startApp (package:flutter_tools/src/commands/daemon.dart:464:12)
<asynchronous suspension>
#30     RunCommand.runCommand (package:flutter_tools/src/commands/run.dart:349:38)
#31     FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:607:18)
#32     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:73:64)
#33     _rootRunUnary (dart:async/zone.dart:1134:38)
#34     _CustomZone.runUnary (dart:async/zone.dart:1031:19)
#35     _FutureListener.handleValue (dart:async/future_impl.dart:139:18)
#36     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:680:45)
#37     Future._propagateToListeners (dart:async/future_impl.dart:709:32)
#38     Future._completeWithValue (dart:async/future_impl.dart:524:5)
#39     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:32:15)
#40     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:290:13)
#41     RunCommand.usageValues (package:flutter_tools/src/commands/run.dart)
#42     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:73:64)
#43     _rootRunUnary (dart:async/zone.dart:1134:38)
#44     _CustomZone.runUnary (dart:async/zone.dart:1031:19)
#45     _FutureListener.handleValue (dart:async/future_impl.dart:139:18)
#46     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:680:45)
#47     Future._propagateToListeners (dart:async/future_impl.dart:709:32)
#48     Future._completeWithValue (dart:async/future_impl.dart:524:5)
#49     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:32:15)
#50     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:290:13)
#51     IosProject.isSwift (package:flutter_tools/src/project.dart)
#52     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:73:64)
#53     _rootRunUnary (dart:async/zone.dart:1134:38)
#54     _CustomZone.runUnary (dart:async/zone.dart:1031:19)
#55     _FutureListener.handleValue (dart:async/future_impl.dart:139:18)
#56     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:680:45)
#57     Future._propagateToListeners (dart:async/future_impl.dart:709:32)
#58     Future._completeWithValue (dart:async/future_impl.dart:524:5)
#59     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:32:15)
#60     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:290:13)
#61     IosProject.buildSettings (package:flutter_tools/src/project.dart)
#62     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:73:64)
#63     _rootRunUnary (dart:async/zone.dart:1134:38)
#64     _CustomZone.runUnary (dart:async/zone.dart:1031:19)
#65     _FutureListener.handleValue (dart:async/future_impl.dart:139:18)
#66     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:680:45)
#67     Future._propagateToListeners (dart:async/future_impl.dart:709:32)
#68     Future._completeWithValue (dart:async/future_impl.dart:524:5)
#69     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:32:15)
#70     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:290:13)
#71     XcodeProjectInterpreter.getBuildSettings (package:flutter_tools/src/ios/xcodeproj.dart)
#72     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:73:64)
#73     _rootRunUnary (dart:async/zone.dart:1134:38)
#74     _CustomZone.runUnary (dart:async/zone.dart:1031:19)
#75     _FutureListener.handleValue (dart:async/future_impl.dart:139:18)
#76     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:680:45)
#77     Future._propagateToListeners (dart:async/future_impl.dart:709:32)
#78     Future._completeWithValue (dart:async/future_impl.dart:524:5)
#79     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:32:15)
#80     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:290:13)
#81     _DefaultProcessUtils.run (package:flutter_tools/src/base/process.dart)
#82     _asyncThenWrapperHelper.<anonymous closure> (dart:async-patch/async_patch.dart:73:64)
#83     _rootRunUnary (dart:async/zone.dart:1134:38)
#84     _CustomZone.runUnary (dart:async/zone.dart:1031:19)
#85     _FutureListener.handleValue (dart:async/future_impl.dart:139:18)
#86     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:680:45)
#87     Future._propagateToListeners (dart:async/future_impl.dart:709:32)
#88     Future._completeWithValue (dart:async/future_impl.dart:524:5)
#89     Future.wait.<anonymous closure> (dart:async/future.dart:400:22)
#90     _rootRunUnary (dart:async/zone.dart:1134:38)
#91     _CustomZone.runUnary (dart:async/zone.dart:1031:19)
#92     _FutureListener.handleValue (dart:async/future_impl.dart:139:18)
#93     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:680:45)
#94     Future._propagateToListeners (dart:async/future_impl.dart:709:32)
#95     Future._complete (dart:async/future_impl.dart:514:7)
#96     _BufferingStreamSubscription.asFuture.<anonymous closure> (dart:async/stream_impl.dart:206:14)
#97     _rootRun (dart:async/zone.dart:1122:38)
#98     _CustomZone.run (dart:async/zone.dart:1023:19)
#99     _CustomZone.runGuarded (dart:async/zone.dart:925:7)
#100    _BufferingStreamSubscription._sendDone.sendDone (dart:async/stream_impl.dart:391:13)
#101    _BufferingStreamSubscription._sendDone (dart:async/stream_impl.dart:401:15)
#102    _BufferingStreamSubscription._close (dart:async/stream_impl.dart:285:7)
#103    _SinkTransformerStreamSubscription._close (dart:async/stream_transformers.dart:98:11)
#104    _EventSinkWrapper.close (dart:async/stream_transformers.dart:25:11)
#105    _StringAdapterSink.close (dart:convert/string_conversion.dart:251:11)
#106    _Utf8ConversionSink.close (dart:convert/string_conversion.dart:302:20)
#107    _ConverterStreamEventSink.close (dart:convert/chunked_conversion.dart:82:18)
#108    _SinkTransformerStreamSubscription._handleDone (dart:async/stream_transformers.dart:143:24)
#109    _rootRun (dart:async/zone.dart:1122:38)
#110    _CustomZone.run (dart:async/zone.dart:1023:19)
#111    _CustomZone.runGuarded (dart:async/zone.dart:925:7)
#112    _BufferingStreamSubscription._sendDone.sendDone (dart:async/stream_impl.dart:391:13)
#113    _BufferingStreamSubscription._sendDone (dart:async/stream_impl.dart:401:15)
#114    _BufferingStreamSubscription._close (dart:async/stream_impl.dart:285:7)
#115    _SyncStreamControllerDispatch._sendDone (dart:async/stream_controller.dart:774:19)
#116    _StreamController._closeUnchecked (dart:async/stream_controller.dart:631:7)
#117    _StreamController.close (dart:async/stream_controller.dart:624:5)
#118    _Socket._onData (dart:io-patch/socket_patch.dart:1838:21)
#119    _rootRunUnary (dart:async/zone.dart:1138:13)
#120    _CustomZone.runUnary (dart:async/zone.dart:1031:19)
#121    _CustomZone.runUnaryGuarded (dart:async/zone.dart:933:7)
#122    _BufferingStreamSubscription._sendData (dart:async/stream_impl.dart:338:11)
#123    _BufferingStreamSubscription._add (dart:async/stream_impl.dart:265:7)
#124    _SyncStreamControllerDispatch._sendData (dart:async/stream_controller.dart:766:19)
#125    _StreamController._add (dart:async/stream_controller.dart:642:7)
#126    _StreamController.add (dart:async/stream_controller.dart:588:5)
#127    new _RawSocket.<anonymous closure> (dart:io-patch/socket_patch.dart:1386:35)
#128    _NativeSocket.issueReadEvent.issue (dart:io-patch/socket_patch.dart:892:18)
#129    _microtaskLoop (dart:async/schedule_microtask.dart:43:21)
#130    _startMicrotaskLoop (dart:async/schedule_microtask.dart:52:5)
#131    _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
#132    _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:175:5)

NoSuchMethodError: The method 'cast' was called on null.
Receiver: null
Tried calling: cast<String>()
Finished with error: Exception: Failed to compile application for the Web.
```

```
Analyzing flutter_developer...                                          

   info • Avoid defining a class that contains only static members • lib/helpers/screenSize.dart:3:1 • avoid_classes_with_only_static_members
   info • The value of the field '_pageBackground' isn't used • lib/main.dart:35:18 • unused_field
   info • The value of the field '_bottomPageBackground' isn't used • lib/main.dart:36:18 • unused_field
   info • Unused import: '../../core/links.dart' • lib/presentation/widgets/help_stepper.dart:6:8 • unused_import
   info • Unused import: '../../core/links.dart' • lib/presentation/widgets/mentoring_stepper.dart:7:8 • unused_import

5 issues found. (ran in 2.2s)

```
```
[✓] Flutter (Channel master, v1.12.3-pre.43, on Mac OS X 10.15.1 19B88, locale en-UA)
    • Flutter version 1.12.3-pre.43 at /Users/mike/Documents/flutter
    • Framework revision 35043cbbe3 (17 hours ago), 2019-11-15 23:52:17 -0500
    • Engine revision 59b5a0cf76
    • Dart version 2.7.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.0)
    • Android SDK at /Users/mike/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.0
    • Java binary at: /Users/mike/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/191.5977832/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.2.1, Build version 11B500
    • CocoaPods version 1.8.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.5)
    • Android Studio at /Users/mike/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/191.5977832/Android Studio.app/Contents
    • Flutter plugin version 41.0.2
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[!] IntelliJ IDEA Ultimate Edition (version 2019.3 BETA)
    • IntelliJ at /Users/mike/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] VS Code (version 1.40.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.6.0

[✓] Connected device (2 available)
    • Chrome     • chrome     • web-javascript • Google Chrome 78.0.3904.97
    • Web Server • web-server • web-javascript • Flutter Tools

! Doctor found issues in 1 category.

```

### PR DESCRIPTION
**Pull request by yjbanov**
## Description

Do not pass missing `dartDefines` as `null`, but as an empty list instead.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/45048

## Tests

Added `--machine` test in `run_test.dart` that uses the daemon.

### CODE REVIEW

**Comment by yjbanov** [comment]
@jonahwilliams It seems `Daemon.daemonCommand` and `Domain.command` were not used at all, so I removed them entirely so they don't cause a similar confusion later. PTAL.

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 1]
```diff
@@ -198,8 +198,16 @@ class AttachCommand extends FlutterCommand {
...
 
     final Daemon daemon = boolArg('machine')
-      ? Daemon(stdinCommandStream, stdoutCommandResponse,
-            notifyingLogger: NotifyingLogger(), logToStdout: true)
+      ? Daemon(
+          stdinCommandStream,
+          stdoutCommandResponse,
+          notifyingLogger: NotifyingLogger(),
+          logToStdout: true,
+          // Since the application was compiled and launched by other means
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/attach.dart]: This comment is a little confusing, its also not quite right, since we can hot restart

Comment (yjbanov) [reply packages/flutter_tools/lib/src/commands/attach.dart]: Should `attach` support `--dart-define` then? It's pretty trivial to add.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/attach.dart]: Yes pls :)

Comment (yjbanov) [reply packages/flutter_tools/lib/src/commands/attach.dart]: Done: https://github.com/flutter/flutter/pull/45317/commits/9d6a11b570ffbb0f933154826865466222a3851d

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**